idaes/core/util/tags.py

```diff
@@ -0,0 +1,432 @@
+        display()"""
+        return self.display(*args, **kwargs)
+
+    def display(self, units=True, format=None, index=None):
+        """Get a string representation of the tagged quantity
+
+        Args:
+            units (bool): Include units of measure in the string
+            format (str): Formatting string, if supplied overrides the default
+            index: If the tagged quantity is indexed, an index for the element
+                to display is required, or the default index is used.
+
+        Returns:
+            str
+        """
+        if self.is_indexed and index is None:
+            index = self._str_index
+        v = self.get_display_value(index=index)
+        if format is None:
+            return self.get_format(units=units, index=index).format(v)
+        else:
+            if units:
+                return self._join_units(format=format, index=index).format(v)
+            else:
+                return format.format(v)
+
+    def _join_units(self, index=None, format=None):
+        """Private method to join the format string with the units of measure
+        string.
+        """
+        if format is None:
+            format = self._format
+        u = self.get_unit_str(index=index)
+        if u == "None" or u == "" or u is None:
+            return format
+        elif u == "%":
+            return "".join([format, u])
+        else:
+            return " ".join([format, u])
+
+    @property
+    def expression(self):
+        """The tagged expression"""
+        return self._expression
+
+    @property
+    def doc(self):
+        """Tag documentation string"""
+        return self._doc
+
+    def get_format(self, units=True, index=None):
+        """Get the formatting string.
+
+        Args:
+            units: if True include units of measure
+            index: index for indexed expressions
+
+        Returns
+            str:
+        """
+        if units:
+            return self._join_units(index=index)
+        else:
+            return self._format
+
+    def get_display_value(self, index=None, convert=True):
+        """Get the value of the expression to display.  Do unit conversion if
+        needed.  This caches the unit conversion, to save time if this is called
+        repeatededly.  The unconverted value is used to ensure the cached
+        converted value is still valid.
+
+        Args:
+            index: index of value to display if expression is indexed
+            convert: if False don't do unit conversion
+
+        Returns:
+            numeric expression value
+        """
+        if self.is_indexed:
+            try:
+                e = self.expression[index]
+            except KeyError:
+                if self._name is None:
+                    raise KeyError(f"{index} is not a valid index for tag")
+                else:
+                    raise KeyError(f"{index} is not a valid index for tag {self._name}")
+        else:
+            e = self.expression
```

Comment:
That's right.  There a default index that gets passed in, whether it is indexed or not from some places.  It's a little weird, but it simplifies dealing with time.  I want to ignore that if it's not indexed.  This will let me group time indexed and not time indexed quantities together.